getinsomnia/insomnia
# Support for gRPC Unary and Client streaming

The following gif shows that you can make unary and client streaming requests to `grpcb.in:9000` using [hello.proto](https://github.com/moul/pb/blob/master/hello/hello.proto).

![2020-11-06 16 33 31](https://user-images.githubusercontent.com/4312346/98324183-76f25e80-2050-11eb-910d-2c18629d688a.gif)

There is no state management in this PR, as that is a significant task on it's own. Therefore, while sending a request, or starting a stream, there is no spinner or feedback. The `stream`, `commit`, and `cancel` buttons should only respond when a stream is running. When a stream is not running, it should show a message in console saying there is no running call. 

Errors (for example, response errors) are _handled_ and written to console. A future PR will improve the error handling and show them in the UI.

I have completed a self-review of the code, but may have missed some things. Please raise any questions/concerns.

## QA:
Try the following workflows with one request in the sidebar:
- Unary ("SayHello"):
  - `send`
    - you should see a response in console
- Client Streaming ("LotsOfGreetings"):
  - `start` -> `stream` -> `stream` -> `commit`
    - you should see a response in console with each message you sent
  - `start` -> `commit`
    - you should see a response in console
  - `start` -> `stream` -> `cancel` 
  - `start` -> `cancel` 
    - you should see an error in console 
  - `stream`
  - `commit`
  - `cancel`
    - you should see a warning in console saying the client call could not be found

An example message body for each RPC is the following. Try this, as well as some invalid message bodies.
```json
{ "greeting": "insomnia" }
```

There is also functionality for concurrently running multiple RPCs, but the state of a request is not visualized in this PR so can be difficult to validate (no spinner shown, etc).

Fixes INS-223, Fixes INS-224, adds functionality for INS-225 and INS-226
